hnn_core/extracellular.py

```diff
@@ -0,0 +1,492 @@
+        assert sigma > 0.0
+        try:  # allow None, but for testing only
+            _validate_type(method, str, 'method')
+            _check_option('method', method, ['psa', 'lsa'])
+        except (TypeError, ValueError) as e:
+            if method is None:
+                pass
+            else:
+                raise e
+        _validate_type(times, list, 'times')
+        _validate_type(voltages, list, 'voltages')
+        for ii in range(len(voltages)):
+            if len(voltages[ii]) != len(positions):
+                raise ValueError(f'number of voltage traces must match number'
+                                 f' of channels, got {len(voltages[ii])} and '
+                                 f'{len(positions)} for trial {ii}')
+            for jj in range(len(voltages[ii])):
+                if len(times) != len(voltages[ii][jj]):
+                    raise ValueError('length of times and voltages must match,'
+                                     f' got {len(times)} and '
+                                     f'{len(voltages[ii][jj])} for trial{ii}, '
+                                     f'channel {jj}')
+
+        self.positions = positions
+        self.sigma = sigma
+        self.method = method
+
+        self._times = times
+        self._data = voltages
+
+    def __getitem__(self, trial_no):
+        try:
+            if isinstance(trial_no, int):
+                return_data = [self._data[trial_no]]
+            elif isinstance(trial_no, slice):
+                return_data = self._data[trial_no]
+            elif isinstance(trial_no, (list, tuple)):
+                return_data = [self._data[trial] for trial in trial_no]
+            else:
+                raise TypeError(f'trial index must be int, slice or list-like,'
+                                f' got: {trial_no} which is {type(trial_no)}')
+        except IndexError:
+            raise IndexError(f'the data contain {len(self)} trials, the '
+                             f'indices provided are out of range: {trial_no}')
+        return ExtracellularArray(self.positions, sigma=self.sigma,
+                                  method=self.method, times=self.times,
+                                  voltages=return_data)
+
+    def __repr__(self):
+        class_name = self.__class__.__name__
+        msg = (f'{len(self.positions)} electrodes, sigma={self.sigma}, '
+               f'method={self.method}')
+        if len(self._data) > 0:
+            msg += f' | {len(self._data)} trials, {len(self.times)} times'
+        return f'<{class_name} | {msg}>'
+
+    def __len__(self):
+        return len(self._data)
+
+    @property
+    def times(self):
+        return self._times
+
+    @property
+    def sfreq(self):
+        """Return the sampling rate of the extracellular data."""
+        dT = np.diff(self.times)
+        Tsamp = np.median(dT)
+        if np.abs(dT.max() - Tsamp) > 1e-3 or np.abs(dT.max() - Tsamp) > 1e-3:
```

Comment:
both the conditions in the `or` looks the same to me ... or is it too late in the night? :)

Reply:
Copy-paste :) It's meant to check that the `max` and `min` are very close to the median.